### PR TITLE
Maximum value for `MultiplierUpdate`

### LINKED ISSUE #12240: Add a maximum value for `trait MultiplierUpdate`
... and enforce it in all implementations, including the one being added in https://github.com/paritytech/substrate/pull/12222

Outcome should be: 

```rust
/// Something that can convert the current multiplier to the next one.
pub trait MultiplierUpdate: Convert<Multiplier, Multiplier> {
	/// Minimum multiplier. Any outcome of the `convert` function should be more than this.
	fn min() -> Multiplier;
	/// Maximum multiplier. Any outcome of the `convert` function should be less than this.
	fn max() -> Multiplier;
	/// Target block saturation level. The `convert` function could possibly update itself based on this target.
	fn target() -> Perquintill;
	/// Variability factor. The `convert function` could possible tune its speed of update based on this.
	fn variability() -> Multiplier;
}
```

This would help implementations ensure that even if the fees should increase forever, they will not go above a certain threshold. 

This remains backwards compatible by setting the maximum to `Bounded::max_value()`.

### PR DESCRIPTION
**Pull request by Szegoo**
Fixes #12240

polkadot companion: https://github.com/paritytech/polkadot/pull/6021

Polkadot address: 126X27SbhrV19mBFawys3ovkyBS87SGfYwtwa8J2FjHrtbmA

### CODE REVIEW

File: frame/transaction-payment/src/lib.rs [hunk 1]
```diff
@@ -217,7 +227,11 @@ where
 
 		if positive {
 			let excess = first_term.saturating_add(second_term).saturating_mul(previous);
-			previous.saturating_add(excess).max(min_multiplier)
+			let next_multiplier = previous.saturating_add(excess).max(min_multiplier);
+			if next_multiplier > X::get() {
+				return X::get()
+			}
+			next_multiplier
```
Comment (notlesh) [inline frame/transaction-payment/src/lib.rs:234]:
```suggestion
			previous
			    .saturating_add(excess)
			    .max(min_multiplier)
			    .min(max_multiplier)
```

I would just chain `.min()` on this. `return X::get()` isn't what you want.

File: frame/transaction-payment/src/lib.rs [hunk 2]
```diff
@@ -233,6 +247,9 @@ impl<M: Get<Multiplier>> MultiplierUpdate for ConstFeeMultiplier<M> {
 	fn min() -> Multiplier {
 		M::get()
 	}
+	fn max() -> Multiplier {
+		<Multiplier as sp_runtime::traits::Bounded>::max_value()
```
Comment (notlesh) [inline frame/transaction-payment/src/lib.rs:251]:
```suggestion
		M::get()
```

This struct should implement `MultiplierUpdate` such that it always returns a constant value. In this case, I think you just want to return `M::get()` like `min()` does.

Comment (Szegoo) [reply frame/transaction-payment/src/lib.rs:250]: Shouldn't this return the maximum value?

**Comment by kianenigma** [comment]
@Szegoo would be good if you link to the issue properly https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue

File: frame/transaction-payment/src/lib.rs [hunk 3]
```diff
@@ -127,12 +127,14 @@ type BalanceOf<T> = <<T as Config>::OnChargeTransaction as OnChargeTransaction<T
 ///
 /// More info can be found at:
 /// <https://research.web3.foundation/en/latest/polkadot/overview/2-token-economics.html>
-pub struct TargetedFeeAdjustment<T, S, V, M>(sp_std::marker::PhantomData<(T, S, V, M)>);
+pub struct TargetedFeeAdjustment<T, S, V, M, X>(sp_std::marker::PhantomData<(T, S, V, M, X)>);
 
 /// Something that can convert the current multiplier to the next one.
 pub trait MultiplierUpdate: Convert<Multiplier, Multiplier> {
 	/// Minimum multiplier
```
Comment (kianenigma) [inline frame/transaction-payment/src/lib.rs:134]:
```suggestion
	/// Minimum multiplier. Any outcome of the `convert` function should be less than this.
```

File: frame/transaction-payment/src/lib.rs [hunk 4]
```diff
@@ -227,12 +237,18 @@ where
 }
 
 /// A struct to make the fee multiplier a constant
-pub struct ConstFeeMultiplier<M: Get<Multiplier>>(sp_std::marker::PhantomData<M>);
+pub struct ConstFeeMultiplier<M: Get<Multiplier>, X: Get<Multiplier>>(
+	sp_std::marker::PhantomData<M>,
+	sp_std::marker::PhantomData<X>,
+);
```
Comment (kianenigma) [inline frame/transaction-payment/src/lib.rs:243]:
```suggestion
pub struct ConstFeeMultiplier<M: Get<Multiplier>, X: Get<Multiplier>>(sp_std::marker::PhantomData<(M, X)>);
```

surprisingly common mistake

File: frame/transaction-payment/src/lib.rs [hunk 5]
```diff
@@ -227,12 +237,18 @@ where
...
 
-impl<M: Get<Multiplier>> MultiplierUpdate for ConstFeeMultiplier<M> {
+impl<M: Get<Multiplier>, X: Get<Multiplier>> MultiplierUpdate for ConstFeeMultiplier<M, X> {
```
Comment (kianenigma) [inline frame/transaction-payment/src/lib.rs:245]: This does not need a separate `X`. The min and max for `ConstFeeMultiplier` are both the same and just `M`.

**Review by kianenigma — APPROVED** [review]
LGTM once my comments are resolved. As usual, please add your polkadot address as well.

**Comment by Szegoo** [comment]
> LGTM once my comments are resolved. As usual, please add your polkadot address as well.

I have resolved the comments. I have added my address, thanks :)

**Comment by Szegoo** [comment]
I have resolved the comments.

I have added my polkadot address, thanks :)

**Comment by notlesh** [comment]
This PR is useful for Moonbeam, we'd love to see it get merged soon :)

File: frame/transaction-payment/src/lib.rs [hunk 6]
```diff
@@ -127,12 +127,14 @@ type BalanceOf<T> = <<T as Config>::OnChargeTransaction as OnChargeTransaction<T
 ///
 /// More info can be found at:
 /// <https://research.web3.foundation/en/latest/polkadot/overview/2-token-economics.html>
-pub struct TargetedFeeAdjustment<T, S, V, M>(sp_std::marker::PhantomData<(T, S, V, M)>);
+pub struct TargetedFeeAdjustment<T, S, V, M, X>(sp_std::marker::PhantomData<(T, S, V, M, X)>);
 
 /// Something that can convert the current multiplier to the next one.
 pub trait MultiplierUpdate: Convert<Multiplier, Multiplier> {
-	/// Minimum multiplier
+	/// Minimum multiplier. Any outcome of the `convert` function should be less than this.
```
Comment (KiChjang) [inline frame/transaction-payment/src/lib.rs:134]: More than or less than?

Comment (Szegoo) [reply frame/transaction-payment/src/lib.rs:134]: More :)

File: frame/transaction-payment/src/lib.rs [hunk 7]
```diff
@@ -217,7 +227,7 @@ where
 
 		if positive {
 			let excess = first_term.saturating_add(second_term).saturating_mul(previous);
-			previous.saturating_add(excess).max(min_multiplier)
+			previous.saturating_add(excess).max(M::get()).min(X::get())
```
Comment (KiChjang) [inline frame/transaction-payment/src/lib.rs:230]: I'd rather put these two in a let binding and use a better identifier:

```suggestion
			previous.saturating_add(excess).max(min_multiplier).min(max_multiplier)
```

**Review by KiChjang — APPROVED** [review]
Looks good after addressing my comment.

File: bin/node-template/runtime/src/lib.rs [hunk 8]
```diff
@@ -262,7 +262,7 @@ impl pallet_transaction_payment::Config for Runtime {
 	type OperationalFeeMultiplier = ConstU8<5>;
 	type WeightToFee = IdentityFee<Balance>;
 	type LengthToFee = IdentityFee<Balance>;
-	type FeeMultiplierUpdate = ConstFeeMultiplier<FeeMultiplier>;
+	type FeeMultiplierUpdate = ConstFeeMultiplier<FeeMultiplierMin>;
```
Comment (ggwpez) [inline bin/node-template/runtime/src/lib.rs:265]:
```suggestion
	type FeeMultiplierUpdate = ConstFeeMultiplier<FeeMultiplier>;
```
It accepts a constant, or?

File: frame/transaction-payment/src/lib.rs [hunk 9]
```diff
@@ -127,12 +127,14 @@ type BalanceOf<T> = <<T as Config>::OnChargeTransaction as OnChargeTransaction<T
 ///
 /// More info can be found at:
 /// <https://research.web3.foundation/en/latest/polkadot/overview/2-token-economics.html>
-pub struct TargetedFeeAdjustment<T, S, V, M>(sp_std::marker::PhantomData<(T, S, V, M)>);
+pub struct TargetedFeeAdjustment<T, S, V, M, X>(sp_std::marker::PhantomData<(T, S, V, M, X)>);
 
 /// Something that can convert the current multiplier to the next one.
 pub trait MultiplierUpdate: Convert<Multiplier, Multiplier> {
-	/// Minimum multiplier
+	/// Minimum multiplier. Any outcome of the `convert` function should be more than this.
```
Comment (ggwpez) [inline frame/transaction-payment/src/lib.rs:134]: *more* or *at least*? Same for `max`.

File: frame/transaction-payment/src/lib.rs [hunk 10]
```diff
@@ -217,7 +228,7 @@ where
 
 		if positive {
 			let excess = first_term.saturating_add(second_term).saturating_mul(previous);
-			previous.saturating_add(excess).max(min_multiplier)
+			previous.saturating_add(excess).max(min_multiplier).min(max_multiplier)
```
Comment (ggwpez) [inline frame/transaction-payment/src/lib.rs:231]: Why is it not added in line 235?

File: bin/node-template/runtime/src/lib.rs (hunk 8 above)
Comment (Szegoo) [reply bin/node-template/runtime/src/lib.rs:265]: Yes, forgot to rename the `FeeMultiplierMin` to `FeeMultiplier`(min and max are the same here)

**Comment by kianenigma** [comment]
bot merge

**Pull request merged**